### PR TITLE
DVGA v1.4.0 (or dare we say v2? :))

### PR DESCRIPTION
**Pull request by plasmak**
Major refactoring, added authentication, authorization, new vulns, removed owners, added moderation feature.

### CODE REVIEW

**Comment by dolevf** [comment]
Hi, 

I wanted to checkout the PR and do a preliminary visual walkthrough of the changes but it seems the Docker instantiation is broken, have you tested your changes using the Dockerfile?

**Comment by plasmak** [comment]
My bad, forgot to fix it. Commit incoming.

**Comment by plasmak** [comment]
@dolevf You'll find authentication info by reading setup.py

**Comment by plasmak** [comment]
TODO : 
- Needs some fixes with jwt verification in hard mode.
- Writing updated solutions.
- Add unit testing.

**Comment by plasmak** [comment]
There's a slightly annoying issue when clicking on the "moderate" or "demoderate" button on a card in "My pastes". It refreshes the page by sending a new qurey for the user's pastes but this query returns the old "public" value. If you hit F5 then it does the query again and gets the correct updated value for the "public" attribute. 
I guess there's a caching mechanisme within graphene and I don't know how to disable it if that's the reason. Will look into it later.

**Comment by plasmak** [comment]
I think the problem comes from the graphene/sqlalchemy implementation (so not a cache thing).
The moderation mutation goes through graphene to change the paste object's public attribute's value while the getPublicPastes query executes plain SQL using the db engine directly. So thi may be a pure sqlalchemy issue.
Looking at this now: https://michaelcho.me/article/sqlalchemy-commit-flush-expire-refresh-merge-whats-the-difference

**Comment by plasmak** [comment]
Yes. It was a sqlalchemy session issue. Calling db.session.flush() fixed the problem.

**Comment by plasmak** [comment]
I know that from a business point of view, the moderator role is not "real life" and does not make a lot of sense. The fact is that the moderator role can moderate other users' posts but cannot demoderate them after (because once moderated they become private to the user). The moderated user cannot demoderate their posts if they don't have the "moderator" role. 
I didn't want the moderator to be able to see all private posts because then this role wouldn't be that different from the admin role. I believe that's not so bad.

File: stuffing.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
+usernames = [
```
Comment (connormckinnon93) [inline stuffing.py:1]: Is there a particular reason to refactor this file in this PR?

I find it makes the PR itself larger than necessary, I would welcome a second PR focused just around the refactor efforts and separate PR for the Authentication and Authorization work.

File: version.py [hunk 2]
```diff
@@ -1 +0,0 @@
-VERSION = '1.3.0'
```
Comment (connormckinnon93) [inline version.py:1]: Similarly, I feel that refactoring could have been it's own PR to improve readability.

File: app.py [hunk 3]
```diff
@@ -1,22 +1,76 @@
...
+from secrets import choice
+
 
 app = Flask(__name__, static_folder="static/")
-app.secret_key = urandom(24)
+app.schema = None
+chrset = string.ascii_lowercase + string.digits
+
+# Weakened deliberately to allow for JWT bruteforce 
+app.secret_key = ''.join([choice(chrset) for _ in range(6)])
```
Comment (connormckinnon93) [inline app.py:24]: I like the idea of including a JWT bruteforce :) in order to improve readability and clarity consider pulling this into a separate PR and updating the README scenarios to include the new vulnerability.

File: app.py [hunk 4]
```diff
@@ -1,22 +1,76 @@
...
+chrset = string.ascii_lowercase + string.digits
+
+# Weakened deliberately to allow for JWT bruteforce 
+app.secret_key = ''.join([choice(chrset) for _ in range(6)]) 
 app.config["SQLALCHEMY_DATABASE_URI"] = config.SQLALCHEMY_DATABASE_URI
 app.config["SQLALCHEMY_TRACK_MODIFICATIONS"] = config.SQLALCHEMY_TRACK_MODIFICATIONS
 app.config["UPLOAD_FOLDER"] = config.WEB_UPLOADDIR
 
 db = SQLAlchemy(app)
+auth = GraphQLAuth(app)
```
Comment (connormckinnon93) [inline app.py:30]: is there a reason that this is defined twice? L30 and L40 I believe...

File: dvga/routes.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+from flask import make_response
+from flask import session
+from flask_graphql import GraphQLView
+
+import config
+from app import app
+from app import db
+from stuffing import solutions as list_of_solutions
+from .decorators import token_required
+from .decorators import access_controlled_for
```
Comment (connormckinnon93) [inline dvga/routes.py:20]: Clever, I like the access controlled for middleware and I like how the authentication and authorization work here.

Did we include any write up in README to indicate the new scenarios?

**Review by connormckinnon93 — CHANGES_REQUESTED** [review]
So I like a lot of the direction that you are going with this but I think a little cleanup first would be appreciated, removing the gitignore has caused a bunch of files to be included that needn't be there.

The refactors are good but distract from the the main thrust of the changes. They could be moved to a separate PR.

JWT bruteforcing and the AuthN/AuthZ scenarios are a welcome addition but I think that they would benefit from inclusion in the README and possible even being two separate PRs.

I have given it a once over but I will try and be more responsive going forward. If you want to ping me after you have reviewed these comments I would be happy to pick it up again and then I can start doing some E2E testings.

File: .gitignore [hunk 6]
```diff
@@ -1,144 +0,0 @@
-# Byte-compiled / optimized / DLL files
```
[the remark this answers is not in the document]
Comment (plasmak) [reply .gitignore:1]: No it was not intentional. I don't know why it was not included as I commited everything in the directory.

File: .vs/ProjectSettings.json [hunk 7]
```diff
@@ -0,0 +1,3 @@
+{
+  "CurrentProjectSetting": null
```
[the remark this answers is not in the document]
Comment (plasmak) [reply .vs/ProjectSettings.json:2]: Yeah, I agree. I thought it was dealt with in the ... .gitignore file: oops.

File: stuffing.py (hunk 1 above)
Comment (plasmak) [reply stuffing.py:1]: I worked locally for a bunch of time before thinking "hey I should contribute it to the project". So I did a lot of work and submitted everything in one PR after reaching out to @dolevf on Twitter. It's the first time I contribute to an open source project so pardon my lack of etiquette please :)

File: version.py (hunk 2 above)
Comment (plasmak) [reply version.py:1]: Ok I get it now.

File: app.py (hunk 3 above)
Comment (plasmak) [reply app.py:24]: I have added multiple vulnerabilities and tried to comment in the code in order to make them visible. Would you like me to do a separate PR for every vulnerability I added or plan to add?

File: app.py (hunk 4 above)
Comment (plasmak) [reply app.py:30]: That's a mistake. I should remove it.

File: dvga/routes.py (hunk 5 above)
Comment (plasmak) [reply dvga/routes.py:20]: Nope. As I have written in the PR comments thread, I did not write the solutions yet. I you also think a paragraph in the README is needed, I'll write it.

**Comment by plasmak** [comment]
@connormckinnon93 Thank you very much for the time spent reviewing this PR. All in all, how do you want me to continue going forward? Should I cancel this PR and resubmit multiple PRs ? If yes, I should start with the refactoring PR and the add the rest of the contributions. Let me know ;)

**Review by plasmak — COMMENTED** [review]
@connormckinnon93 Thank you very much for the time spent reviewing this PR. All in all, how do you want me to continue going forward? Should I cancel this PR and resubmit multiple PRs ? If yes, I should start with the refactoring PR and the add the rest of the contributions. Let me know ;)

**Comment by dolevf** [comment]
I would recommend the following:

* PR for refactoring
* PR for only the authentication and authorization layer with documentation
* PR for each vulnerability, with associated solutions and README updates (x 2 if there are 2 new vulns)

This PR is becoming very heavy, I suggest to maybe close and start a new clean one with our proposals.

**Comment by plasmak** [comment]
> 
> 
> I would recommend the following:
> 
>     * PR for refactoring
> 
>     * PR for only the authentication and authorization layer with documentation
> 
>     * PR for each vulnerability, with associated solutions and README updates (x 2 if there are 2 new vulns)
> 
> 
> This PR is becoming very heavy, I suggest to maybe close and start clean ones for each bullet.

Well noted. I'll do that this week end or earlier if can spare some time.

**Pull request closed without merging**